### ISSUE TITLE
Connection leaking?

### ISSUE THREAD

**Issue by xorander00** [requery/requery#506]
I'm using requery on a server-side project and it's mostly working great.  However, after a number of queries, it throws a SQL exception complaining that the connection pool is exhausted.

Do I have to explicitly release a connection somehow after a query?  I thought about wrapping queries with a try-with-resources block, but it doesn't look like it supports AutoClosable.  Here's an example of one of my queries that crashes after a number of times being called:

```
return _dbStoreEntity.runInTransaction(() -> {
            return _dbStoreEntity.select(MyModel.class).where(
                    MyModel.USER_NAME.equal(userName))
                    .and(MyModel.USER_ID.equal(userId))
                    .get().firstOrNull();
        });
```

It wasn't originally wrapped in a transaction, but I saw #436 and figured I'd give it a shot.  I also tried changing from .firstOrNull() to .first(), but that didn't make a difference.

Please correct me if I'm being an idiot :)

**Comment by xorander00** [comment]
Oh, and _dbStoreEntity is EntityDataStore<Persistable>.  Also, it's in a singleton class since I'm using this with Dropwizard/Jersey and resources are treated as Singletons by default.  I wouldn't think it would be an issue though, but maybe I'm wrong.

**Comment by xorander00** [comment]
Update: This is when using SQLite via the [Xerial sqlite-jdbc](https://github.com/xerial/sqlite-jdbc) driver.  I haven't tested it yet against either MSSQL or PostgreSQL.  It's creating the DataSource from [Dropwizard's DataSourceFactory](http://www.dropwizard.io/1.1.0/dropwizard-db/apidocs/io/dropwizard/db/DataSourceFactory.html), although I wouldn't think that would make a difference.

I did tweak the settings on the DataSourceFactory & made validation & eviction settings aggressive as well as setting timeouts low and checking connections on create/borrow/idle/return.  It's working now, but the driver is throwing warnings about connections being abandoned (as expected).  It's at least killing the stale connections though and creating new ones.

Not sure what's holding onto the connections and not returning them to the pool.

**Comment by npurushe** [comment]
Hi the `Result` type does implement the `Closeable` interface and can be closed with a try-with-resources block. However terminating methods like `firstOrNull` will close the connection. Are you sure it's now another query that is causing the problem?

**Comment by xorander00** [comment]
It's possible, I'm not sure.  I have one call that uses a Result<T>, which I'm wrapping with try-with-resources block.  Going to see if this is the issue.  I'm hoping it is, since it's an easy fix.

**Comment by xorander00** [comment]
Nope, still getting it :/ The code is pretty straightforward, nothing complex.  It's literally just processing a GET /api/mycall incoming request by issuing two queries:

_dbStoreEntity.select(MyModel.class).where(
                    MyModel.USER_NAME.equal(userName))
                    .and(MyModel.USER_ID.equal(userId))
                    .get().firstOrNull();

The resulting model has a field which is then used to issue another query, which gets a Result<T>:

_dbStoreEntity.select(RowsModel.class).where(
                    RowsModel.USER_ID.equal(userId))
                    .get();

I originally tried to pull all of the records in with a single join query, but had some trouble figuring out how to set up the relational models.  I shelved that for now and am just using two separate queries, but will come back to it later.

Any ideas on what could be keeping the connection(s) open?  I was looking through my code to see if anything could be holding a reference outside of scope/lifecycle, but so far I don't see anything.  Are the requery methods thread-safe?

**Comment by xorander00** [comment]
Ok yeah, just confirmed that all queries are holding onto the connection and not returning them to the pool, not just some of them.

So terminal queries like firstOrNull are supposed to return the connection to the pool, correct?  Although it shouldn't be, I wonder if this is specific to the SQLite driver?

**Comment by npurushe** [comment]
Yes terminating operations close the JDBC connection. You can verify that by placing a breakpoint in ResultSetIterator.close() where it closes the connection and see if that is hit.

**Comment by xorander00** [comment]
Yup, it's being hit.  I haven't had the chance to look into it deeply yet, but in the quick debugging session I just ran through, I think I saw that closed was set to true even though the JDBC connection was still open.  Upon a subsequent request, I could see the pool size growing with active connections that had released = false set.

I'm not 100% sure what's going on yet, but I'll take a further look in the coming few days.  If you have any suggestions on what else I should look into, let me know :)  I'll let you know what I find.

**Comment by spouzols** [comment]
Hello, not sure if this gives you more leads or closes some, but I was having the same kind of issue in a Spring Boot app when using the default Tomcat jdbc pool and clearly each query was eating a connection never released to the pool and was seeing the same exhaustion you describe, gave Hikari a try as it's what we usually use on other apps, and it seems to be working okay - but that may well be a coping mechanism in Hikari making it work around an actual issue.

**Comment by xorander00** [comment]
Yeah, I'm guessing it's a configuration/usage issue.  Still, I'd like to leave this issue open for a little bit longer to figure it out, since it'll probably be helpful for future requery users that are coming from app/api framework corner.

@spouzols What's the DAO/ORM that Spring Boot offers by default?  I wonder if we can intersect JDBC configuration between that and Dropwizard to get an idea.

**Comment by spouzols** [comment]
By default Spring Boot points to Hibernate directly or Hibernate as JPA implementation, starters dependencies usually grab Hibernate as implementation, and the Tomcat JDBC pool also by default.

**Comment by xorander00** [comment]
Heads up.  Switching to directly creating and using a DataSource seems to have fixed the problem.  I'm not sure what the issue was, but my guess is something specific to the underlying configuration and/or the metrics/reporting stuff that's integrated (for me anyway at least, with Dropwizard and all).

**Comment by npurushe** [comment]
This should probably be resolved by #533, which deals specifically with pooled connections, looks like all the connections are closed when using `Result` along with try with resources, or the normal entity api methods

**Comment by alex-shpak** [comment]
Came here from google. If anyone will hit this issue.
Problem is that, when releasing connection, requery uses `statement.getConnection()` which returns original connection (for example `PgConnection` rather than `ProxyConnection` from tomcat pool.
Pool expects that `ProxyConnection.close()` be called, releasing connection, which is not happening
